ext/git.c

```diff
@@ -115,104 +137,88 @@ void removeCredentials(zend_string* repo_url) {
-    return add_git_info(meta, get_DD_GIT_COMMIT_SHA(), get_DD_GIT_REPOSITORY_URL(), is_root_span, true);
+bool inject_from_global_tags(zval *carrier) {
+    zend_array * global_tags = get_DD_TAGS();
+    zval *commit_sha = zend_hash_str_find(global_tags, ZEND_STRL("git.commit.sha"));
+    zval *repository_url = zend_hash_str_find(global_tags, ZEND_STRL("git.repository_url"));
+    if (commit_sha && repository_url && Z_TYPE_P(commit_sha) == IS_STRING && Z_TYPE_P(repository_url) == IS_STRING) {
+        remove_credentials(Z_STR_P(repository_url));
+        add_git_info(carrier, Z_STR_P(commit_sha), Z_STR_P(repository_url));
+        return true;
+    }
+
+    return false;
 }
 
-bool inject_from_global_tags(zval* meta, bool is_root_span) {
-    zend_array* global_tags = get_DD_TAGS();
+bool inject_from_git_dir(zval *carrier, zend_string* cwd) {
+    zend_string *git_dir = find_git_dir(ZSTR_VAL(cwd));
+    zend_string *commit_sha = get_commit_sha(ZSTR_VAL(git_dir));
+    zend_string *repository_url = get_repository_url(ZSTR_VAL(git_dir));
     bool success = false;
 
-    if (global_tags) {
-        zval* git_commit_sha = zend_hash_str_find(global_tags, ZEND_STRL("git.commit.sha"));
-        zval* git_repository_url = zend_hash_str_find(global_tags, ZEND_STRL("git.repository_url"));
-
-        if (git_commit_sha && git_repository_url && Z_TYPE_P(git_commit_sha) == IS_STRING &&
-            Z_TYPE_P(git_repository_url) == IS_STRING) {
-            success = add_git_info(meta, Z_STR_P(git_commit_sha), Z_STR_P(git_repository_url), is_root_span, true);
-        }
-    }
-
-    if (success && is_root_span) {
-        zend_hash_str_del(Z_ARR_P(meta), ZEND_STRL("git.commit.sha"));
-        zend_hash_str_del(Z_ARR_P(meta), ZEND_STRL("git.repository_url"));
+    if (commit_sha && repository_url && ZSTR_LEN(commit_sha) > 0 && ZSTR_LEN(repository_url) > 0) {
+        remove_credentials(repository_url);
+        add_git_info(carrier, commit_sha, repository_url);
+        success = true;
+        zend_string_release(git_dir);
     }
 
     return success;
 }
 
-bool inject_from_git_files(zval* meta, bool is_root_span) {
-    char cwd[PATH_MAX];
-    if (!getcwd(cwd, sizeof(cwd))) {
-        return false;
-    }
-
-    char git_dir[PATH_MAX];
-    snprintf(git_dir, sizeof(git_dir), "%s/.git", cwd);
-
-    zend_string* git_commit_sha = get_commit_sha(git_dir);
-    zend_string* git_repository_url = get_repository_url(git_dir);
+void ddtrace_inject_git_metadata(zval *git_metadata_zv) {
+    zend_string *cwd_zstr = NULL;
 
-    if (!git_commit_sha || !git_repository_url) {
-        if (git_commit_sha) zend_string_release(git_commit_sha);
-        if (git_repository_url) zend_string_release(git_repository_url);
-        return false;
+    if (SG(options) & SAPI_OPTION_NO_CHDIR) {
+        const char *script_filename = SG(request_info).path_translated;
+        cwd_zstr = zend_string_init(script_filename, strlen(script_filename), 1);
```

Comment:
path_translated is the full path, not just the dir, you'll need to strip the filename.